python/lsst/daf/butler/_butler.py

```diff
@@ -297,17 +270,44 @@ def from_config(
         arguments provided, but it defaults to `False` when there are not
         collection arguments.
         """
-        cls = cls._find_butler_class(config, searchPaths)
-        return cls(
-            config,
-            collections=collections,
-            run=run,
-            searchPaths=searchPaths,
-            writeable=writeable,
-            inferDefaults=inferDefaults,
-            **kwargs,
+        # DirectButler used to have a way to specify a "copy constructor" by
+        # passing the "butler" parameter to its constructor.  This
+        # functionality has been moved out of the constructor into
+        # Butler._clone(), but the new interface is not public yet.
+        butler = kwargs.pop("butler", None)
+        if butler is not None:
+            if not isinstance(butler, Butler):
+                raise TypeError("'butler' parameter must be a Butler instance")
+            if config is not None or searchPaths is not None or writeable is not None:
+                raise TypeError(
+                    "Cannot pass 'config', 'searchPaths', or 'writeable' arguments with 'butler' argument."
+                )
+            return butler._clone(collections=collections, run=run, inferDefaults=inferDefaults, **kwargs)
+
+        options = ButlerInstanceOptions(
+            collections=collections, run=run, writeable=writeable, inferDefaults=inferDefaults, kwargs=kwargs
         )
 
+        # Load the Butler configuration.  This may involve searching the
+        # environment to locate a configuration file.
+        butler_config = ButlerConfig(config, searchPaths=searchPaths, without_datastore=without_datastore)
+        butler_type = butler_config.get_butler_type()
+
+        # Make DirectButler if class is not specified.
+        if butler_type == ButlerType.DIRECT:
+            from .direct_butler import DirectButler
+
+            return DirectButler.create_from_config(
+                butler_config,
+                options=options,
+                without_datastore=without_datastore,
+            )
+        elif butler_type == ButlerType.REMOTE:
+            from .remote_butler import RemoteButlerFactory
+
+            factory = RemoteButlerFactory.create_factory_from_config(butler_config)
+            return factory.create_butler_with_credentials_from_environment(butler_options=options)
+
```

Comment:
Because mypy's exhaustiveness checking will ensure that all the cases are handled, there's some value in not having the raise.  Without the raise, when someone adds an enum value, mypy will tell them all the places that need to do something with it.

I don't feel that strongly about it one way or the other though.

Reply:
`match` naturally comes with a fall through clause that can raise. I feel better if the code looks like it's dealing with all the options without relying on mypy.